assembly/src/library.rs

```diff
@@ -0,0 +1,434 @@
+        &self.version
+    }
+
+    fn modules(&self) -> Self::ModuleIterator<'_> {
+        self.modules.iter()
+    }
+}
+
+impl LibraryContents {
+    // CONSTANTS
+    // --------------------------------------------------------------------------------------------
+
+    /// File extension for the Miden Assembly Library.
+    pub const FILE_EXTENSION: &str = "masl";
+
+    // SERIALIZATION HELPERS
+    // --------------------------------------------------------------------------------------------
+
+    /// Write the namespace, version and modules to a target.
+    fn write_contents_into<I, IB>(
+        namespace: &LibraryNamespace,
+        version: &Version,
+        mut modules: I,
+        target: &mut ByteWriter,
+    ) -> Result<(), SerializationError>
+    where
+        I: Iterator<Item = IB> + ExactSizeIterator,
+        IB: Borrow<Module>,
+    {
+        namespace.write_into(target)?;
+        version.write_into(target)?;
+        target.write_len(modules.len())?;
+        modules.try_for_each(|module| {
+            let module = module.borrow();
+            ModulePath::strip_namespace(&module.path).write_into(target)?;
+            module.ast.write_into(target)?;
+            Ok(())
+        })
+    }
+
+    /// Write a library into the target. This is separated from the library trait because it
+    /// introduces a new requirement that the iterator of modules is `ExactSizeIterator`.
+    pub fn write_library_into<L>(
+        library: &L,
+        target: &mut ByteWriter,
+    ) -> Result<(), SerializationError>
+    where
+        L: Library,
+        for<'a> <L as Library>::ModuleIterator<'a>: ExactSizeIterator,
+    {
+        Self::write_contents_into(
+            library.root_ns(),
+            library.version(),
+            library.modules(),
+            target,
+        )
+    }
+}
+
+impl Serializable for LibraryContents {
+    fn write_into(&self, target: &mut ByteWriter) -> Result<(), SerializationError> {
+        Self::write_contents_into(&self.namespace, &self.version, self.modules.iter(), target)
+    }
+}
+
+impl Deserializable for LibraryContents {
+    fn read_from(bytes: &mut ByteReader) -> Result<Self, SerializationError> {
+        let namespace = LibraryNamespace::read_from(bytes)?;
+        let version = Version::read_from(bytes)?;
+
+        let len = bytes.read_len()?;
+        let modules = (0..len)
+            .map(|_| {
+                ModulePath::read_from(bytes)
+                    .map(|path| path.to_absolute(&namespace))
+                    .and_then(|path| ModuleAst::read_from(bytes).map(|ast| (path, ast)))
+                    .map(|(path, ast)| Module { path, ast })
+            })
+            .collect::<Result<_, _>>()?;
+
+        Ok(Self {
+            namespace,
+            version,
+            modules,
+        })
+    }
+}
+
+#[cfg(feature = "std")]
+mod use_std {
+    use super::*;
+    use std::{collections::HashSet, fs, io, path::Path};
+
+    impl LibraryContents {
```

Comment:
I was actually thinking of a slightly different interface here. Something more like:

```Rust
pub fn read_from_dir<P>(path: P) -> io::Result<Self>
pub fn write_to_file<P>(&self, path: P) -> io::Result<()>
```
The first one would read a library from a directory containing `.masm` files (basically, very similar to what is happening in `build.rs` of `stdlib`.

The second one would just write out the library as a single `.masl` file.

I do think that having a function for reading a set of libraries from a directory could be useful. But I'm not sure if there is a use case for writing a set of libraries to a directory. So, I'd probably keep the `read_dir()` (although maybe a different name is needed), but I would probably get rid of `write_to_dir()` with `write_to_file()`.

Reply:
Refactored with a small change on the output function: we just take a dir, and we create a masl with the namespace as file name.

We can quickly revert this change if it is not desirable.